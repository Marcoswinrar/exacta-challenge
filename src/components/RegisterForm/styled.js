import styled from 'styled-components'
import media from 'styled-media-query'

export const RegisterForm = styled.div`
`
export const Container = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 1.2rem;

  ${media.lessThan("large")`
    flex-direction: column;  
  `}
`
