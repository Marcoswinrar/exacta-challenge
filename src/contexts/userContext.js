import { createContext, useContext, useEffect, useState } from 'react'
import UserModal from '../components/UserModal'
import { ToastContainer, toast } from 'react-toastify'

const UsersContext = createContext({
  users: [],
  editUser: (id) => null,
  removeUser: (id) => null,
  notify: (message, type) => null
})

const UsersProvider = ({ children }) => {
  const [showModal, setShowModal] = useState(false)
  const [formDefaultValues, setFormDefaultValues] = useState({})
  const [users, setUsers] = useState([])

  const notify = (message, type) => toast[type](message)

  const editUser = (user) => {
    const newUsers = [...users]
    newUsers[user.index] = user
    setUsers(newUsers)
  }

  const removeUser = (id) => {
    const filtered = users.filter((user) => user.id !== id)
    setUsers([...filtered])
  }

  useEffect(() => {
    setFormDefaultValues({})
  }, [])

  return (
    <UsersContext.Provider
      value={{
        users,
        setUsers,
        setShowModal,
        setFormDefaultValues,
        editUser,
        removeUser,
        notify
      }}
    >
      {children}
      {showModal && <UserModal initialValues={formDefaultValues} />}
      <ToastContainer position="bottom-right" />
    </UsersContext.Provider>
  )
}

const userContext = () => useContext(UsersContext)
export { UsersProvider, userContext }
