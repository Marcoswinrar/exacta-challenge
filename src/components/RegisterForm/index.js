import { useUsersContext } from '../../contexts/userContext'
import Form from '../Form'
import Button from '../Button'
import Fetch from '../../utils/fetch'
import * as S from './styled'

const RegisterForm = () => {
  const { notify } = useUsersContext()

  const handleSubmit = async (user) => {
    try {
      await Fetch.post('users', user)
      notify("Usu√°rio criado com sucesso!", "success")

    } catch (error) {
      if (error.response) {
        notify("Houve um erro ao cadastrar!", "error")
      }
    }

  }

  return (
    <S.RegisterForm>
      <Form onSubmit={handleSubmit}>
        <Button width="200px" type="submit" value="Cadastrar" />
      </Form>
    </S.RegisterForm>
  )
}

export default RegisterForm