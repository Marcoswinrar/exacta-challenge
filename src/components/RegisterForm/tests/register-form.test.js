import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import { act } from "react-dom/test-utils"
import RegisterForm from ".."
import { UsersProvider } from "../../../contexts/userContext"
import Fetch from "../../../utils/fetch"

describe("Componente <RegisterForm/>", () => {

  beforeAll(() => {
    jest.spyOn(Fetch, "post")
      .mockImplementation(() => Promise.resolve({}))
  })

  beforeEach(() => {
    render(
      <RegisterForm />
      , { wrapper: UsersProvider }
    )
  })

  it("Deve renderizar o Cadastro sem problemas", () => {

    expect(screen.queryByText(/Cadastrar/)).toBeInTheDocument()
    expect(screen.getByRole("button")).toHaveStyle({
      "width": "200px"
    })
  })

  describe("Teste de envio de cadastro", () => {

    it("Deve cadastrar o usuário e apresentar mensagem de sucesso", async () => {

      const nome = screen.getByLabelText("Nome:")
      const numeroRg = screen.getByLabelText("Numero do RG:")
      const dataEmissao = screen.getByLabelText("Data Emissão:")
      const orgaoEmissor = screen.getByLabelText("Orgão Emissor:")
      const genero = screen.getByLabelText("Gênero:")

      const botao = screen.getByRole("button")

      await act(async () => {
        fireEvent.change(nome, { target: { value: "Marcos" } })
        fireEvent.change(numeroRg, { target: { value: "12.345.678-9" } })
        fireEvent.change(dataEmissao, { target: { value: "20/12/2013" } })
        fireEvent.change(orgaoEmissor, { target: { value: "SSP" } })
        fireEvent.change(genero, { target: { value: "M" } })

        fireEvent.submit(botao)
      })

      await waitFor(() => {
        expect(screen.queryByText(/Usuário criado com sucesso!/i)).toBeInTheDocument()
      })

    })

    it("Deve falhar ao cadastrar e apresentar mensagem de erro", async () => {

      jest.spyOn(Fetch, "post")
        .mockImplementation(() => Promise.reject({ response: "404 Not found" }))

      const nome = screen.getByLabelText("Nome:")
      const numeroRg = screen.getByLabelText("Numero do RG:")
      const dataEmissao = screen.getByLabelText("Data Emissão:")
      const orgaoEmissor = screen.getByLabelText("Orgão Emissor:")
      const genero = screen.getByLabelText("Gênero:")

      const botao = screen.getByRole("button")

      await act(async () => {
        fireEvent.change(nome, { target: { value: "Marcos" } })
        fireEvent.change(numeroRg, { target: { value: "12.345.678-9" } })
        fireEvent.change(dataEmissao, { target: { value: "20/12/2013" } })
        fireEvent.change(orgaoEmissor, { target: { value: "SSP" } })
        fireEvent.change(genero, { target: { value: "M" } })

        fireEvent.submit(botao)
      })

      await waitFor(() => {
        expect(screen.queryByText(/Houve um erro ao cadastrar!/i)).toBeInTheDocument()
      })

    })
  })
})