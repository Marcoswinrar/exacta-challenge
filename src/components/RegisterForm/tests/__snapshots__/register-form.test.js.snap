// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente <RegisterForm/> Teste de envio de cadastro Deve cadastrar o usuário e apresentar mensagem de sucesso 1`] = `
<div>
  <div
    class="sc-evZas kdknvw"
  >
    <form
      class="sc-jqUVSM jgNWSl"
    >
      <div
        class="sc-papXJ iHWkgh"
      >
        <div
          class="sc-kDDrLX cRFGXq"
        >
          <label
            class="sc-iBkjds hTxmns"
            for="Nome"
          >
            Nome:
          </label>
          <input
            class="sc-gKXOVf gllsEC"
            id="Nome"
            name="Nome"
          />
        </div>
        <div
          class="sc-kDDrLX cRFGXq"
        >
          <label
            class="sc-iBkjds hTxmns"
            for="NumeroRg"
          >
            Numero do RG:
          </label>
          <input
            id="NumeroRg"
            value=""
          />
        </div>
        <div
          class="sc-kDDrLX cRFGXq"
        >
          <label
            class="sc-iBkjds hTxmns"
            for="DataEmissao"
          >
            Data Emissão:
          </label>
          <input
            id="DataEmissao"
            value=""
          />
        </div>
        <div
          class="sc-kDDrLX cRFGXq"
        >
          <label
            class="sc-iBkjds hTxmns"
            for="OrgaoEmissor"
          >
            Orgão Emissor:
          </label>
          <select
            id="OrgaoEmissor"
            name="OrgaoEmissor"
          >
            <option
              value=""
            >
              Selecione
            </option>
            <option
              value="SSP"
            >
              Secretaria de Segurança Pública
            </option>
            <option
              value="PM"
            >
              Polícia Militar
            </option>
            <option
              value="PC"
            >
              Policia Civil
            </option>
            <option
              value="CNT"
            >
              Carteira Nacional de Habilitação
            </option>
            <option
              value="DIC"
            >
              Diretoria de Identificação Civil
            </option>
            <option
              value="CTPS"
            >
              Carteira de Trabaho e Previdência Social
            </option>
            <option
              value="FGTS"
            >
              Fundo de Garantia do Tempo de Serviço
            </option>
            <option
              value="IFP"
            >
              Instituto Félix Pacheco
            </option>
            <option
              value="IPF"
            >
              Instituto Pereira Faustino
            </option>
            <option
              value="IML"
            >
              Instituto Médico-Legal
            </option>
            <option
              value="MTE"
            >
              Ministério do Trabalho e Emprego
            </option>
            <option
              value="MMA"
            >
              Ministério da Marinha
            </option>
            <option
              value="MAE"
            >
              Ministério da Aeronáutica
            </option>
            <option
              value="MEX"
            >
              Ministério do Exército
            </option>
            <option
              value="POF"
            >
              Polícia Federal
            </option>
            <option
              value="POM"
            >
              Polícia Militar
            </option>
            <option
              value="SES"
            >
              Carteira de Estrangeiro
            </option>
            <option
              value="SJS"
            >
              Secretaria da Justiça e Segurança
            </option>
            <option
              value="SJTS"
            >
              Secretaria da Justiça do Trabalho e Segurança
            </option>
            <option
              value="ZZZ"
            >
              Outros (inclusive exterior)
            </option>
          </select>
        </div>
        <div
          class="sc-kDDrLX cRFGXq"
        >
          <label
            class="sc-iBkjds hTxmns"
            for="Genero"
          >
            Gênero:
          </label>
          <select
            id="Genero"
            name="Genero"
          >
            <option
              value=""
            >
              Selecione
            </option>
            <option
              value="M"
            >
              Masculino
            </option>
            <option
              value="F"
            >
              Feminino
            </option>
          </select>
        </div>
      </div>
      <button
        class="sc-iqcoie ehrxCs"
        type="submit"
        width="200px"
      >
        Cadastrar
      </button>
    </form>
  </div>
</div>
`;
